Mac OS Compatibility: pthread_tryjoin_np

Thanks for open sourcing this library. It looks like great code and I'm eager to try it out!

I tried compiling on Mac but ran into some issues. There were some minor problems with the configure script and one compilation error. Unfortunately `pthread_threadid_np` is not supported on Mac OS.

```
FcgiDecodingApp.cc:233:10: error: use of undeclared identifier
      'pthread_tryjoin_np'; did you mean 'pthread_threadid_np'?
                                if (!pthread_tryjoin_np(*i, NULL)) {
                                     ^~~~~~~~~~~~~~~~~~
                                     pthread_threadid_np
```

--- a/configure
+++ b/configure
@@ -26,6 +26,7 @@
 CONFIGURE_VERSION=1
 OUTPUT_MK=apiai.mk
 INCLUDE_PATHS="/usr/include /usr/local/include"
+LIBRARY_PATHS="/usr/lib /usr/local/lib /usr/local/lib64"
 
 APIAI_CXX_FLAGS=
 
@@ -104,15 +105,19 @@ function exit_success {
 }
 
 function check_sys_library {
-	echo -n "Looking for $1 library: "
-	local response=$(whereis lib$1)
-	local libpath=${response##*:}
-	if [ -z "$libpath" ]; then
-		echo "Not found"
-		failure "Library $1 not found"
-	else
-		echo $libpath
-	fi
+  for file in $1; do
+    for path in $LIBRARY_PATHS; do
+      local result="${path}/lib${file}"
+      echo -n "Checking ${result}..." >&2
+      if [ -f "$result" ]; then
+        echo "OK" >&2
+        echo $result
+        break 2
+      else
+        echo "Not found" >&2
+      fi
+    done
+  done
 }
 
 function check_header_file {
@@ -136,8 +141,8 @@ echo "Configuring ..."
 echo "Looking for Kaldi sources in \"$KALDIROOT\"..."
 check_exists "$KALDIROOT/src/kaldi.mk"
 
-check_sys_library fcgi
-check_sys_library fcgi++
+check_sys_library fcgi.a
+check_sys_library fcgi++.a
 
 FCGIO_H=$(check_header_file "fcgio.h")
 if [ -z "$FCGIO_H" ]; then

--- a/configure
+++ b/configure
@@ -31,8 +31,14 @@ APIAI_CXX_FLAGS=
 
 function rel2abs {
   if [ ! -z "$1" ]; then
-    local retval=$(readlink -f $1)
-    echo $retval
+    case "${1}" in
+      [./]*)
+      echo "$(cd ${1%/*}; pwd)/${1##*/}"
+      ;;
+      *)
+      echo "${PWD}/${1}"
+      ;;
+    esac
   fi
 }
 
@@ -60,7 +66,7 @@ if ! [ -x "$PWD/configure" ]; then
 fi
 
 ## Default locations for Kaldi sources.
-KALDIROOT=$(readlink -f ../kaldi-apiai)
+KALDIROOT=$(rel2abs ../kaldi-apiai)
 
 function usage {
   echo 'Usage: ./configure [--kaldi-root=KALDIROOT]';

--- a/src/FcgiDecodingApp.cc
+++ b/src/FcgiDecodingApp.cc
@@ -227,15 +227,12 @@ int FcgiDecodingApp::Run(int argc, char **argv) {
 
 		KALDI_VLOG(1) << "Threads ready: " << thread_list.size();
 
-		while (thread_list.size() > 0) {
-			sleep(1);
-			for (std::list<pthread_t>::iterator i = thread_list.begin(); i != thread_list.end(); i++) {
-				if (!pthread_tryjoin_np(*i, NULL)) {
-					thread_list.erase(i);
-					KALDI_VLOG(1) << "Thread finished, threads left: " << thread_list.size();
-				}
+		for (std::list<pthread_t>::iterator i = thread_list.begin(); i != thread_list.end(); ++i) {
+			if ((errnumber = pthread_join(*i, NULL)) != 0) {
+				KALDI_WARN << "Failed to join thread: " << strerror(errnumber);
 			}
 		}
+		KALDI_VLOG(1) << "Thread finished, threads left: " << thread_list.size();
 	}
 
 	running_ = false;

